### ISSUE TITLE
LoRa + Multi GPUs: RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one.

### ISSUE THREAD

**Issue by mahdip72** [huggingface/peft#951]
### System Info

```
accelerate==0.22.0
peft==0.5.0
transformers==4.33.1
```
platform: ubuntu 22.04.3 LTS
python: 3.10.12

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

I am using models from ESM-V2 family such as "facebook/esm2_t6_8M_UR50D". 

```python
self.model = EsmModel.from_pretrained(model_name)

config = LoraConfig(
    r=8,
    lora_alpha=32,
    target_modules=[
        "dense"
                    ],
    inference_mode=False,
    lora_dropout=0.05,
    bias="none",
 )
self.model = get_peft_model(self.model, config)
```

When I use lora on multiple GPUs using accelerate, I got the following error:
```
Steps:   0%|                                                                                                                                                                                                          | 0/28 [00:00<?, ?it/s]/home/mpngf/environments/joint_training/lib/python3.10/site-packages/torch/nn/functional.py:4999: UserWarning: Support for mismatched key_padding_mask and attn_mask is deprecated. Use same type for both instead.
  warnings.warn(
/home/mpngf/environments/joint_training/lib/python3.10/site-packages/torch/nn/functional.py:4999: UserWarning: Support for mismatched key_padding_mask and attn_mask is deprecated. Use same type for both instead.
  warnings.warn(
Steps:   4%|██████▌                                                                                                                                                                               | 1/28 [00:02<01:01,  2.29s/it, lr=1.02e-5]Traceback (most recent call last):
  File "/home/mpngf/projects/JointTraining/train.py", line 326, in <module>
    main(config_file)
  File "/home/mpngf/projects/JointTraining/train.py", line 287, in main
    train_loss, train_acc = train(epoch, accelerator, tools, global_step)
  File "/home/mpngf/projects/JointTraining/train.py", line 90, in train
    preds = tools['net'](batch)
  File "/home/mpngf/environments/joint_training/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/mpngf/environments/joint_training/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 1139, in forward
    if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by
making sure all `forward` function outputs participate in calculating loss.
If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
Parameter indices which did not receive grad for rank 0: 36 37
 In addition, you can set the environment variable TORCH_DISTRIBUTED_DEBUG to either INFO or DETAIL to print out information about which particular parameters did not receive gradient on this rank as part of this error
Traceback (most recent call last):
  File "/home/mpngf/projects/JointTraining/train.py", line 326, in <module>
    main(config_file)
  File "/home/mpngf/projects/JointTraining/train.py", line 287, in main
    train_loss, train_acc = train(epoch, accelerator, tools, global_step)
  File "/home/mpngf/projects/JointTraining/train.py", line 90, in train
    preds = tools['net'](batch)
  File "/home/mpngf/environments/joint_training/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/mpngf/environments/joint_training/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 1139, in forward
    if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by
making sure all `forward` function outputs participate in calculating loss.
If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
Parameter indices which did not receive grad for rank 1: 36 37
 In addition, you can set the environment variable TORCH_DISTRIBUTED_DEBUG to either INFO or DETAIL to print out information about which particular parameters did not receive gradient on this rank as part of this error
``` 
But on single GPU there is not a problem and it works fine!

### Expected behavior

I expect the LoRa method to work on both single and multi GPUs. But it shows an error when using multiple GPUs.

**Comment by BenjaminBossan** [comment]
This is being addressed in #905. If you have time, you could give that branch a try and report back if it solves your issue.

**Comment by mahdip72** [comment]
@BenjaminBossan I remove the official peft and install that branch using this command:
```
pip install git+https://github.com/BenjaminBossan/peft.git@fix-setting-requires-grad#egg=peft
```

Still I am getting the same error on multi GPUs.

**Comment by mahdip72** [comment]
Also, need to mention that w/o using LoRa, my training code works fine on multiple GPUs.

**Comment by BenjaminBossan** [comment]
Thanks for testing. Strange that the issue remains. Could you please ensure that the installed PEFT version is indeed `0.6.0dev0`?

Also, could you please follow the instruction from the error message so that we can see which parameters are causing the issue?

> In addition, you can set the environment variable TORCH_DISTRIBUTED_DEBUG to either INFO or DETAIL to print out information about which particular parameters did not receive gradient on this rank as part of this error

i.e. re-run your script with `TORCH_DISTRIBUTED_DEBUG=DETAIL`.

**Comment by mahdip72** [comment]
@BenjaminBossan I checked that and the version of PEFT is `0.60dev0`.
I re-run my training code and I got this:
```
Parameters which did not receive grad for rank 0: encoder.model.base_model.model.pooler.dense.lora_B.default.weight, encoder.model.base_model.model.pooler.dense.lora_A.default.weight
Parameter indices which did not receive grad for rank 0: 72 73
```

**Comment by BenjaminBossan** [comment]
Thanks for the update. Is the `"pooler"` layer part of the `modules_to_save` in your config?

**Comment by mahdip72** [comment]
I did that and I got the following error:
```
Parameters which did not receive grad for rank 0: encoder.model.base_model.model.pooler.modules_to_save.default.dense.lora_B.default.weight, encoder.model.base_model.model.pooler.modules_to_save.default.dense.lora_A.default.weight, encoder.model.base_model.model.pooler.modules_to_save.default.dense.bias, encoder.model.base_model.model.pooler.modules_to_save.default.dense.weight
Parameter indices which did not receive grad for rank 0: 36 37 38 39
```

**Comment by BenjaminBossan** [comment]
Hmm I see. It seems that this pooler layer does not participate in the loss computation, but it has `requires_grad = True`. Could you please undo what I suggested above (adding it to `modules_to_save`). Next you need to either manually select this module and call `requires_grad_(False)` on it, or you have to edit the `target_modules` argument in your lora config so that it does _not_ target the pooler (which has "dense" in part of it's name) but still all other target layers. This could be achieved via some form of regex.

To explain, since the pooler's name is matching with `target_modules`, PEFT applies LoRA on it and sets `requires_grad=True`. However, this layer is not part of the computation of the loss. But PyTorch DDP doesn't like to have parameters with `requires_grad` set to `True` when they don't add to the loss. Therefore, you need to prevent this specific layer to have gradients. Why it is not part of the loss computation, I don't know.

**Comment by mahdip72** [comment]
Thanks @BenjaminBossan! This works for me:

```python
from peft import LoraConfig, get_peft_model

model = EsmModel.from_pretrained(facebook/esm2_t6_8M_UR50D)
config = LoraConfig(
    r=8,
    lora_alpha=32,
    target_modules=[
        "dense"
                    ],
    inference_mode=False,
    lora_dropout=0.1,
    bias="none",
)
self.model = get_peft_model(model, config)

for param in model.pooler.parameters():
    param.requires_grad = False

```

Is there a bug we have to report? How can we fix it without manually setting the gradient to False?

**Comment by BenjaminBossan** [comment]
It's not really a bug IMO, it's rather a weird interaction.

1. There is a model with a layer that is not being used for the loss calculation (pooler)
2. By default, PEFT sets `requires_grad` to `False` on this layer but it is "accidentally" part of the `target_modules`, so `requires_grad` is set to `True`.
3. DDP does not like params with `requires_grad=True` but not participating in loss calculation (for reasons unclear to me), so raises an error.

This is just an unfortunate chain of events and not something we can provide a general fix for in PEFT.

Note that the "proper" solution would be not to apply LoRA to the pooler layer. But for your specific situation, the solution you posted is easier than fiddling with the `target_modules` argument, so it should be okay.

**Comment by mahdip72** [comment]
@BenjaminBossan, when I activate LoRa for my ESM model as described above, I've noticed that the GPU memory usage is significantly higher, even though the number of trainable parameters is roughly the same compared to when fine-tuning the last layer. This observation holds true under the same training configurations. I find this unusual. For further clarification, I've also checked with 4-bit quantization, and the memory usage does not change. Could you please advise if this is expected behavior?

**Comment by BenjaminBossan** [comment]
Since it's a different problem, could you please open a new issue? Please include exactly what you're doing, the hardware setting (e.g. only multi GPU or same on single GPU?) and the memory numbers you observe? Thanks. Also, feel free to close this issue if your initial question was answered.

**Comment by jinlHe** [comment]
I got exactly the same error as the  author @mahdip72 .
> RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by
> making sure all `forward` function outputs participate in calculating loss.
> If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
> Parameter indices which did not receive grad for rank 1: 36 37
>  In addition, you can set the environment variable TORCH_DISTRIBUTED_DEBUG to either INFO or DETAIL to print out information about which particular parameters did not receive gradient on this rank as part of this error
> ```
> 
> But on single GPU there is not a problem and it works fine!
> 
> ### Expected behavior
> I expect the LoRa method to work on both single and multi GPUs. But it shows an error when using multiple GPUs.

I tried lora on the open source [Mamba-based cobra](https://github.com/h-zhao1997/cobra)(an open source VLM trained on [mamba-2.8b-zephyr](https://huggingface.co/xiuyul/mamba-2.8b-zephyr)), loaded with mamba-2.8b-zephyr weights.
Here's the code I used to initialize the llm:
`    def init_llm(cls, llm_backbone_id, llm_max_length, low_resource=False, low_res_device=0, lora_r=0,
                 lora_target_modules=["x_proj", "embeddings", "in_proj", "out_proj"], **lora_kargs):
        logging.info('Loading Mamba')

        mamba_model = MambaLLMBackbone(
            llm_backbone_id=llm_backbone_id,
            llm_max_length=llm_max_length
        )

        mamba_tokenizer = mamba_model.get_tokenizer()
        if lora_r > 0:
            mamba_model = prepare_model_for_kbit_training(mamba_model)
            lora_config = LoraConfig(
                r=lora_r,
                bias="none",
                task_type="CAUSAL_LM",
                target_modules=lora_target_modules,
                **lora_kargs
            )
            mamba_model = get_peft_model(mamba_model, lora_config)
            mamba_model.print_trainable_parameters()
            # for name, param in mamba_model.named_parameters():
            #     param.requires_grad = False
                # if any(module in name for module in lora_target_modules):
                #     param.requires_grad = True
        else:
            for name, param in mamba_model.named_parameters():
                param.requires_grad = False

        # for name, param in mamba_model.named_parameters():
        #     if param.requires_grad:
        #         print(name)
        logging.info('Loading Mamba Done')
        return mamba_model, mamba_tokenizer`

However, after loss.backward(), the grad of all lora layers is None, which also leads to this problem, and I have not found the reason:
`            with torch.cuda.amp.autocast(enabled=use_amp):
                loss, loss_dict = self.train_step(model=model, samples=samples)
                loss /= accum_grad_iters #TODO: not affect loss_dict values for logging

            # after_train_step()
            if use_amp:
                scaler.scale(loss).backward()
            else:
                loss.backward()

            for name, param in model.named_parameters():
                if param.requires_grad and param.grad is None:
                    print(f"{name} requires_grad and not used forward pass.")`
Print (part of it) :
`
mamba_model.base_model.model.llm.backbone.layers.0.mixer.in_proj.lora_A.default.weight requires_grad and not used forward pass.
mamba_model.base_model.model.llm.backbone.layers.0.mixer.in_proj.lora_B.default.weight requires_grad and not used forward pass.
mamba_model.base_model.model.llm.backbone.layers.0.mixer.x_proj.lora_A.default.weight requires_grad and not used forward pass.
mamba_model.base_model.model.llm.backbone.layers.0.mixer.x_proj.lora_B.default.weight requires_grad and not used forward pass.
mamba_model.base_model.model.llm.backbone.layers.0.mixer.out_proj.lora_A.default.weight requires_grad and not used forward pass.
mamba_model.base_model.model.llm.backbone.layers.0.mixer.out_proj.lora_B.default.weight requires_grad and not used forward pass.
mamba_model.base_model.model.llm.backbone.layers.1.mixer.in_proj.lora_A.default.weight requires_grad and not used forward pass.
mamba_model.base_model.model.llm.backbone.layers.1.mixer.in_proj.lora_B.default.weight requires_grad and not used forward pass.
mamba_model.base_model.model.llm.backbone.layers.1.mixer.x_proj.lora_A.default.weight requires_grad and not used forward pass.
mamba_model.base_model.model.llm.backbone.layers.1.mixer.x_proj.lora_B.default.weight requires_grad and not used forward pass.
mamba_model.base_model.model.llm.backbone.layers.1.mixer.out_proj.lora_A.default.weight requires_grad and not used forward pass.
mamba_model.base_model.model.llm.backbone.layers.1.mixer.out_proj.lora_B.default.weight requires_grad and not used forward pass.
`
I checked the whole model, including the visual part, by printing param.requires_grad (the model is similar to cobra, it is a visual encoder plus mamba LLM). param.requires_grad of the whole model is OK. However, the lora part appears grad=None.

Manually setting requires_grad=False as the authors did won't throw an error, but it freezes the whole mamba LLM and makes LoRA useless.
[mamba's official documentation](https://huggingface.co/state-spaces/mamba-2.8b-hf) indicates that LoRA can be used, but I didn't use LoRA the way they gave me. But now the problem with the code also makes me very confused, I don't know what the problem is.I would appreciate it if someone could help me solve this problem.
